engine/consensus/matching/core.go

```diff
@@ -209,35 +229,45 @@ func (c *Core) onReceipt(originID flow.Identifier, receipt *flow.ExecutionReceip
 	//  => drop Receipt
 	sealed, err := c.state.Sealed().Head()
 	if err != nil {
-		return fmt.Errorf("could not find sealed block: %w", err)
+		return false, fmt.Errorf("could not find sealed block: %w", err)
 	}
 
 	isSealed := head.Height <= sealed.Height
 	if isSealed {
 		log.Debug().Msg("discarding receipt for already sealed and finalized block height")
-		return nil
+		return false, nil
 	}
 
-	// if the receipt is not valid, because the parent result is unknown, we will drop this receipt.
-	// in the case where a child receipt is dropped because it is received before its parent receipt, we will
-	// eventually request the parent receipt with the `requestPending` function
 	childSpan := c.tracer.StartSpanFromParent(receiptSpan, trace.CONMatchOnReceiptVal)
 	err = c.receiptValidator.Validate([]*flow.ExecutionReceipt{receipt})
 	childSpan.Finish()
+
+	if validation.IsMissingPreviousResultError(err) {
+		// If previous result is missing, we can't validate this receipt.
+		// althrough we will request its previous (potentially multiple receipts),
+		// We don't want to drop it now, because when the missing previous arrive
+		// in a wrong order, they will still be dropped, and causing the catch up
+		// to be inefficient.
+		// Instead, we could cache the receipt in case it arrives earlier than its
+		// previous receipt.
+		// for instance, given blocks A <- B <- C <- D <- E, if we receive their receipts
+		// in the order of [E,C,D,B,A], then:
+		// if we drop the missing previous receipts, then only A will be processed;
+		// if we cache the missing previous receipts, then all of them will be processed, because
+		// once A is processed, we will check if there is a child receipt pending,
+		// if yes, then process it.
+		c.pendingReceipts.Add(receipt)
+		log.Info().Msg("receipt is cached because its previous result is missing")
+		return false, nil
+	}
+
 	if err != nil {
-		if engine.IsInvalidInputError(err) {
-			log.Err(err).Msg("invalid execution receipt")
-			return nil
-		}
-		return fmt.Errorf("failed to validate execution receipt: %w", err)
+		return false, fmt.Errorf("failed to validate execution receipt: %w", err)
 	}
 
-	err = c.storeReceipt(receipt, head, &log)
+	_, err = c.storeReceipt(receipt, head)
 	if err != nil {
-		if engine.IsDuplicatedEntryError(err) {
-			return nil
-		}
-		return fmt.Errorf("failed to store receipt: %w", err)
+		return false, fmt.Errorf("failed to store receipt: %w", err)
```

Comment:
Very good question here.

The problem is that we are storing receipts and incorporated results in two different mempools. And we are making an assumption that if a receipt already exist in the receipts mempool, then it must have been stored in incorporated results, and we can exit without storing it in the incorporated results mempool.

This is a dangerous assumption. Because the two different mempools have different ejection rules, and it's possible that the two mempool could get out of sync. I want to be safe to just try storing to the incorporated results mempool again, even if the receipt is a duplicate. And since the operations are all in memory, the addtional storing will be trivial, but the increased safety is valuable.